node/src/components/consensus/protocols/highway.rs

```diff
@@ -582,22 +584,33 @@ where
                     }
                 }
             }
+            Ok(HighwayMessage::RequestLatestState(panorama)) => {
+                let state = self.highway.state();
+
+                let merged_panorama = state.panorama().merge(state, &panorama);
```

Comment:
Isn't `panorama` what our peer sees, and `state.panorama()` what we see?
We can't merge those: We can only merge panoramas that we know. This will panic if `panorama` contains any entries we don't have.

What we want to send them is not the merged panorama, but all those entries in `state.panorama()` that are different from `panorama`. If they have a _later_ unit from some validater, then we have no way of knowing that, based on the hash they sent us.

So maybe it's not worth including their panorama here after all?

Reply:
I was wondering about that, but it actually still works!

What `Panorama::merge` does when we don't have an observation included in the other panorama, is it marks the corresponding validator as faulty. That would be wrong if we used this knowledge to actually mark the validator as faulty in the state... but we only used the `Correct` observations from the merged panorama. So the end result is exactly what you say we want!

Reply:
Right… except if the validator equivocated and we have one fork and the peer has the other fork… then we assume that their unit is later, and they assume that ours is later? We should probably actually request their unit from them in that case?

Reply:
Also, let's make sure that behavior is well-documented in `merge`, and not just accidental. :sweat_smile:

Reply:
I'm pretty sure it's not accidental, because if we have all the units referred to in the panoramas, this case points to an equivocation :wink: 

> except if the validator equivocated and we have one fork and the peer has the other fork… then we assume that their unit is later, and they assume that ours is later?

Yes, pretty much. In that case right now we would hope that this gets resolved by units being broadcast later :sweat_smile: 

> We should probably actually request their unit from them in that case?

That sounds like a good idea, but it looks like this calls for another request type, and it might be outside the scope of this PR. I can try implementing that, though, if you think it's worth it.

Reply:
Wouldn't it just be "requesting a dependency"? Even though it's not really a dependency of anything else.

Anyway, I don't think it's a blocker.

Reply:
I thought that we would zip both panoramas and iterate over each observation pair, compare them with `geq` method and return units for which the local panorama has a greater observation :thinking:

Reply:
> Wouldn't it just be "requesting a dependency"? Even though it's not really a dependency of anything else

Hmm, you may be right.

> I thought that we would zip both panoramas and iterate over each observation pair, compare them with is_greater (like geq but without the q part) method and return units for which the local panorama has a greater observation

That's basically what `Panorama::merge` does :wink: